File: baekjoon/gold/1016/1016.py
# input
# 1 1000
# output
# 608

# 4, 9, 16, 25, 36, 49, 

import math


n, m = map(int, input().split())
prime = [1 for _ in range(int(m ** 0.5) + 1)]
prime[0], prime[1] = 0, 0
square = []
for i in range(2, len(prime)):
  if prime[i] == 1:
    for j in range(i + i, len(prime), i):
      prime[j] = 0
    square.append(i ** 2)

r = [1 for _ in range(m - n + 1)]

for i in square:
  start_num = math.ceil(n / i) * i # 이걸 생각 못함
  for j in range(start_num, m + 1, i):
    r[j - n] = 0

print(sum(r))

File: baekjoon/gold/1016/1016.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var n, m int

	fmt.Scan(&n, &m)
	r := initValue[int](make([]int, m-n+1), 1)

	for _, v := range primeSquare(m) {
		startNum := int(math.Ceil(float64(n)/float64(v))) * v
		for j := startNum; j <= m; j += v {
			r[j-n] = 0
		}
	}
	ans := 0
	for _, v := range r {
		if v == 1 {
			ans++
		}
	}
	fmt.Println(ans)
}

func primeSquare(max int) []int {
	var square []int
	prime := initValue[int](make([]int, int(math.Sqrt(float64(max)))+1), 1)
	prime[0], prime[1] = 0, 0
	for i := 2; i < len(prime); i++ {
		if prime[i] == 1 {
			square = append(square, i*i)
			for j := i + i; j < len(prime); j += i {
				prime[j] = 0
			}
		}
	}
	return square
}

func initValue[T any](t []T, v T) []T {
	for i := range t {
		t[i] = v
	}
	return t
}
